promexporter/prometheus.go

```diff
@@ -105,3 +107,39 @@ func (c customGatherer) Gather() ([]*dto.MetricFamily, error) {
 	c.getMetricsFn()
 	return c.oldGatherer.Gather()
 }
+
+// GetLabelNames creates a list of label names out of struct fields to use in Prometheus metric
+// The label can be specified by a `label` tag, e.g.:
+// ```go
+// type ProcessLabels struct {
+// 	ProcessName   string `label:"process_name"`
+// 	ProcessStatus string `label:"process_status"`
+// }
+// ```
+// If no `label` tag is specified, a field name converted to snake_case will be used instead
+func GetLabelNames(labelStruct interface{}) []string {
+	t := reflect.TypeOf(labelStruct)
+	labels := make([]string, t.NumField())
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		tag := f.Tag.Get("label")
+		if tag == "" {
+			tag = strcase.ToSnake(f.Name)
+		}
+		labels[i] = tag
+	}
+	return labels
+}
+
+// GetLabelValues creates a list of label values out of struct field values to use in Prometheus metric
+// See GetLabelNames function for the context
+// If a field type is not string it will be converted to string automatically, see https://golang.org/pkg/reflect/#Value.String
+func GetLabelValues(labelStruct interface{}) []string {
+	v := reflect.ValueOf(labelStruct)
+	labels := make([]string, v.NumField())
+	for i := 0; i < v.NumField(); i++ {
+		f := v.Field(i)
+		labels[i] = f.String()
```

Comment:
What happens when a field is not exposed?

Reply:
It doesn't really have any effect. I added a test to verify that. Seems like reflection is able to extract this information just fine.